operators/audio/audio_decoder.hpp

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "ocos.h"
+
+#include <list>
+#define DR_FLAC_IMPLEMENTATION
+#include "dr_flac.h"
+#define DR_MP3_IMPLEMENTATION 1
+#define DR_MP3_FLOAT_OUTPUT 1
+#include "dr_mp3.h"
+#define DR_WAV_IMPLEMENTATION
+#include "dr_wav.h"
+
+#include "string_utils.h"
+#include "string_tensor.h"
+
+struct KernelAudioDecoder : public BaseKernel {
+ public:
+  KernelAudioDecoder(const OrtApi& api, const OrtKernelInfo& info) : BaseKernel(api, info) {
+  }
+
+  enum {
+    DEFAULT_STREAM = 0,
+    WAV_STREAM,
+    MP3_STREAM,
+    FLAC_STREAM
+  };
+
+  int ReadStreamFormat(OrtKernelContext* context, const int64_t* p_data) {
+    int stream_format = DEFAULT_STREAM;
+    const OrtValue* ov_format = ort_.KernelContext_GetInput(context, 1);
+    if (ov_format != nullptr) {
+      std::vector<std::string> str_format;
+      GetTensorMutableDataString(api_, ort_, context, ov_format, str_format);
+      const char* formats[] = {"default",
+                               "wav",
+                               "mp3",
+                               "flac"};
+      for (auto fmt : formats) {
+        if (str_format[0] == fmt) {
+          break;
+        }
+        stream_format++;
+      }
+    }
+
+    if (stream_format == DEFAULT_STREAM){
+      auto p_stream = reinterpret_cast<char const*>(p_data);
+      std::string_view marker(p_stream, 4);
+      if (marker == "fLaC") {
+        stream_format = FLAC_STREAM;
+      } else if (marker == "RIFF") {
+        stream_format = WAV_STREAM;
+      } else if (marker[0] == char(-1) && marker[1] == char(-13)) {
+        stream_format = MP3_STREAM;
+      } else {
+        ORTX_CXX_API_THROW("[AudioDecoder]: Unknown audio stream format", ORT_INVALID_ARGUMENT);
+      }
+    }
+    return stream_format;
+  }
+
+  template<typename _CLS_AUDIO, typename _FN_DECODER>
+  static size_t DrReadFrames(std::list<std::vector<float>>& frames, _FN_DECODER fx, _CLS_AUDIO& obj) {
+    const size_t default_chunk_size = 4096;
+    int64_t total_frames = 0;
+
+    for (;;) {
+      std::vector<float> buf;
+      buf.resize(default_chunk_size * obj.channels);
+      auto n_frames = fx(&obj, default_chunk_size, buf.data());
+      if (n_frames <= 0) {
+        break;
+      }
+      auto n_samples = n_frames * obj.channels;
+      total_frames += n_samples;
```

Comment:
I'm confused by the names here. why are we incrementing the frame count by a number of samples?

Reply:
"frame" and "sample" seem to be separate units. is this not the case?